url4fdf79ae2c_beckchr_juel_tgz-pJ8-de_odysseus_el_samples_xml_sax_XMELFilterJ8/src/de/odysseus/el/tree/Tree.java

```diff
@@ -0,0 +1,153 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */ 
+package de.odysseus.el.tree;
+
+import java.lang.reflect.Method;
+import java.util.List;
+
+import javax.el.ELException;
+import javax.el.FunctionMapper;
+import javax.el.ValueExpression;
+import javax.el.VariableMapper;
+
+import de.odysseus.el.misc.LocalMessages;
+import de.odysseus.el.misc.TypeConverter;
+
+/**
+ * Parsed expression, usually created by a {@link de.odysseus.el.tree.TreeBuilder}.
+ * The {@link #bind(FunctionMapper, VariableMapper)} method is used to create
+ * {@link de.odysseus.el.tree.Bindings}, which are needed at evaluation time to
+ * lookup functions and variables. The tree itself does not contain such information,
+ * because it would make the tree depend on the function/variable mapper supplied at
+ * parse time.
+ * 
+ * @author Christoph Beck
+ */
+public class Tree {
+	private final ExpressionNode root;
+	private final List<FunctionNode> functions;
+	private final List<IdentifierNode> identifiers;
+	private final boolean deferred;
+
+	/**
+	 * 
+	 * Constructor.
+	 * @param root root node
+	 * @param functions collection of function nodes
+	 * @param identifiers collection of identifier nodes
+	 */
+	public Tree(ExpressionNode root, List<FunctionNode> functions, List<IdentifierNode> identifiers, boolean deferred) {
+		super();
+		this.root = root;
+		this.functions = functions;
+		this.identifiers = identifiers;
+		this.deferred = deferred;
+	}
+
+	/**
+	 * Get function nodes (in no particular order)
+	 */
+	public Iterable<FunctionNode> getFunctionNodes() {
+		return functions;
+	}
+	
+	/**
+	 * Get identifier nodes (in no particular order)
+	 */
+	public Iterable<IdentifierNode> getIdentifierNodes() {
+		return identifiers;
+	}
+	
+	/**
+	 * @return root node
+	 */
+	public ExpressionNode getRoot() {
+		return root;
+	}
+	
+	public boolean isDeferred() {
+		return deferred;
+	}
+	
+	@Override
+	public String toString() {
+		return getRoot().getStructuralId(null);
+	}
+
+	/**
+	 * Create a bindings.
+	 * @param fnMapper the function mapper to use
+	 * @param varMapper the variable mapper to use
+	 * @return tree bindings
+	 */
+	public Bindings bind(FunctionMapper fnMapper, VariableMapper varMapper) {
+		return bind(fnMapper, varMapper, null);
+	}
+
+	/**
+	 * Create a bindings.
+	 * @param fnMapper the function mapper to use
+	 * @param varMapper the variable mapper to use
+	 * @param converter custom type converter
+	 * @return tree bindings
+	 */
+	public Bindings bind(FunctionMapper fnMapper, VariableMapper varMapper, TypeConverter converter) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 13. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 139-151 into a separate method.